providers/structured/provider.go

```diff
@@ -47,16 +50,25 @@ func (p StructuredOutputLogProvider) GetRecordEventCalls() []RecordEventCallArgs
 
 // NewStructuredOutputLogProvider returns a LogProvider which records all calls made to it.
 func NewStructuredOutputLogProvider() *StructuredOutputLogProvider {
+	return LogProvider(nil).(*StructuredOutputLogProvider)
+}
+
+func LogProvider(nextProvider providers.LogProvider) providers.LogProvider {
 	rawLogCalls := map[providers.RawLogCallType][]RawLogCallArgs{}
 	for _, callType := range providers.RawLogCallTypes() {
 		rawLogCalls[callType] = []RawLogCallArgs{}
 	}
 
-	return &StructuredOutputLogProvider{
+	lp := &StructuredOutputLogProvider{
 		rawLogCalls:      rawLogCalls,
 		recordCalls:      []RecordCallArgs{},
 		recordEventCalls: []RecordEventCallArgs{},
 	}
+
+	if nextProvider != nil {
+		lp.LogProvider = chaining.LogProvider(nextProvider)
```

Comment:
What you want to do is here always set lp.LogProvider to the chaining provider, even if nextProvider is nil; then you don't need to bother with all the nil-checking in individual methods, because the chaining provider does it for you (that's its main job).

Reply:
Ah ok.. Didn't check the internals of the chaining provider.